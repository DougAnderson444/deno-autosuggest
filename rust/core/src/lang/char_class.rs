use std::fmt;
use super::Lang;


pub trait CharPattern: fmt::Debug {
    fn matches(&self, ch: char, lang: &Option<Lang>) -> Option<bool>;
}


#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum CharClass {
    Any,
    Control,
    Whitespace,
    Punctuation,
    NotAlphaNum,
    Consonant,
    Vowel,
}

use CharClass::{
    Any,
    Control,
    Whitespace,
    Punctuation,
    NotAlphaNum,
    Consonant,
    Vowel,
};


impl CharPattern for CharClass {
    fn matches(&self, ch: char, lang: &Option<Lang>) -> Option<bool> {
        match self {
            Any         => Some(true),
            Control     => Some(ch.is_control()),
            Whitespace  => Some(ch.is_whitespace()),
            Punctuation => Some(ch.is_ascii_punctuation()),
            NotAlphaNum => Some(!ch.is_alphanumeric()),
            Consonant   => Some(lang.as_ref()?.get_char_class(ch)? == Consonant),
            Vowel       => Some(lang.as_ref()?.get_char_class(ch)? == Vowel),
        }
    }
}


impl<P: CharPattern> CharPattern for [P] {
    fn matches(&self, ch: char, lang: &Option<Lang>) -> Option<bool> {
        let mut met_none = false;
        for pattern in self {
            match pattern.matches(ch, lang) {
                Some(true)  => return Some(true),
                Some(false) => continue,
                None => {
                    met_none = true;
                    continue;
                },
            }
        }
        if met_none { None } else { Some(false) }
    }
}


impl<P: CharPattern> CharPattern for [P; 1] { fn matches(&self, ch: char, lang: &Option<Lang>) -> Option<bool> { self[..].matches(ch, lang) } }
impl<P: CharPattern> CharPattern for [P; 2] { fn matches(&self, ch: char, lang: &Option<Lang>) -> Option<bool> { self[..].matches(ch, lang) } }
impl<P: CharPattern> CharPattern for [P; 3] { fn matches(&self, ch: char, lang: &Option<Lang>) -> Option<bool> { self[..].matches(ch, lang) } }
impl<P: CharPattern> CharPattern for [P; 4] { fn matches(&self, ch: char, lang: &Option<Lang>) -> Option<bool> { self[..].matches(ch, lang) } }
impl<P: CharPattern> CharPattern for [P; 5] { fn matches(&self, ch: char, lang: &Option<Lang>) -> Option<bool> { self[..].matches(ch, lang) } }


#[cfg(test)]
mod tests {
    use super::{CharClass, CharPattern, Lang};

    use CharClass::{
        Any,
        Control,
        Whitespace,
        Punctuation,
        NotAlphaNum,
        Consonant,
        Vowel,
    };

    #[test]
    fn pattern_matches_any() {
        assert_eq!(Any.matches('\0', &None), Some(true));
        assert_eq!(Any.matches('2',  &None), Some(true));
        assert_eq!(Any.matches('f',  &None), Some(true));
        assert_eq!(Any.matches(';',  &None), Some(true));
        assert_eq!(Any.matches(' ',  &None), Some(true));
    }

    #[test]
    fn pattern_matches_control() {
        assert_eq!(Control.matches('\0', &None), Some(true));
        assert_eq!(Control.matches('2',  &None), Some(false));
        assert_eq!(Control.matches('f',  &None), Some(false));
        assert_eq!(Control.matches(';',  &None), Some(false));
        assert_eq!(Control.matches(' ',  &None), Some(false));
    }

    #[test]
    fn pattern_matches_whitespaces() {
        assert_eq!(Whitespace.matches(' ',  &None), Some(true));
        assert_eq!(Whitespace.matches('\t', &None), Some(true));
        assert_eq!(Whitespace.matches('\n', &None), Some(true));
        assert_eq!(Whitespace.matches('2',  &None), Some(false));
        assert_eq!(Whitespace.matches('f',  &None), Some(false));
        assert_eq!(Whitespace.matches(';',  &None), Some(false));
        assert_eq!(Whitespace.matches('\0', &None), Some(false));
    }

    #[test]
    fn pattern_matches_punctuation() {
        assert_eq!(Punctuation.matches(';',  &None), Some(true));
        assert_eq!(Punctuation.matches('.',  &None), Some(true));
        assert_eq!(Punctuation.matches(',',  &None), Some(true));
        assert_eq!(Punctuation.matches('2',  &None), Some(false));
        assert_eq!(Punctuation.matches('f',  &None), Some(false));
        assert_eq!(Punctuation.matches(' ',  &None), Some(false));
        assert_eq!(Punctuation.matches('\0', &None), Some(false));
    }

    #[test]
    fn pattern_matches_non_alphanum() {
        assert_eq!(NotAlphaNum.matches('2',  &None), Some(false));
        assert_eq!(NotAlphaNum.matches('f',  &None), Some(false));
        assert_eq!(NotAlphaNum.matches('ы',  &None), Some(false));
        assert_eq!(NotAlphaNum.matches('も', &None), Some(false));
        assert_eq!(NotAlphaNum.matches(';',  &None), Some(true));
        assert_eq!(NotAlphaNum.matches('.',  &None), Some(true));
        assert_eq!(NotAlphaNum.matches(',',  &None), Some(true));
        assert_eq!(NotAlphaNum.matches(' ',  &None), Some(true));
        assert_eq!(NotAlphaNum.matches('\0', &None), Some(true));
    }

    #[test]
    fn pattern_matches_slice() {
        let pattern = &[Whitespace, Punctuation][..];

        assert_eq!(pattern.matches(' ',  &None), Some(true));
        assert_eq!(pattern.matches(';',  &None), Some(true));
        assert_eq!(pattern.matches('f',  &None), Some(false));
        assert_eq!(pattern.matches('2',  &None), Some(false));
        assert_eq!(pattern.matches('\0', &None), Some(false));
    }

    #[test]
    fn pattern_matches_array_2() {
        let pattern = [Whitespace, Punctuation];

        assert_eq!(pattern.matches(' ',  &None), Some(true));
        assert_eq!(pattern.matches(';',  &None), Some(true));
        assert_eq!(pattern.matches('f',  &None), Some(false));
        assert_eq!(pattern.matches('2',  &None), Some(false));
        assert_eq!(pattern.matches('\0', &None), Some(false));
    }

    #[test]
    fn pattern_matches_vowel_no_lang() {
        assert_eq!(Vowel.matches('2',  &None), None);
        assert_eq!(Vowel.matches('f',  &None), None);
        assert_eq!(Vowel.matches('ы',  &None), None);
        assert_eq!(Vowel.matches('も', &None), None);
        assert_eq!(Vowel.matches(';',  &None), None);
        assert_eq!(Vowel.matches('.',  &None), None);
        assert_eq!(Vowel.matches(',',  &None), None);
        assert_eq!(Vowel.matches(' ',  &None), None);
        assert_eq!(Vowel.matches('\0', &None), None);
    }

    #[test]
    fn pattern_matches_vowel_with_lang() {
        let mut lang = Lang::new();
        lang.add_char_class('f', Consonant);
        lang.add_char_class('ы', Vowel);
        let lang = Some(lang);

        assert_eq!(Vowel.matches('2',  &lang), None);
        assert_eq!(Vowel.matches('f',  &lang), Some(false));
        assert_eq!(Vowel.matches('ы',  &lang), Some(true));
        assert_eq!(Vowel.matches('も', &lang), None);
        assert_eq!(Vowel.matches(';',  &lang), None);
        assert_eq!(Vowel.matches('.',  &lang), None);
        assert_eq!(Vowel.matches(',',  &lang), None);
        assert_eq!(Vowel.matches(' ',  &lang), None);
        assert_eq!(Vowel.matches('\0', &lang), None);
    }

    #[test]
    fn pattern_matches_consonant_no_lang() {
        assert_eq!(Consonant.matches('2',  &None), None);
        assert_eq!(Consonant.matches('f',  &None), None);
        assert_eq!(Consonant.matches('ы',  &None), None);
        assert_eq!(Consonant.matches('も', &None), None);
        assert_eq!(Consonant.matches(';',  &None), None);
        assert_eq!(Consonant.matches('.',  &None), None);
        assert_eq!(Consonant.matches(',',  &None), None);
        assert_eq!(Consonant.matches(' ',  &None), None);
        assert_eq!(Consonant.matches('\0', &None), None);
    }

    #[test]
    fn pattern_matches_consonant_with_lang() {
        let mut lang = Lang::new();
        lang.add_char_class('f', Consonant);
        lang.add_char_class('ы', Vowel);
        let lang = Some(lang);

        assert_eq!(Consonant.matches('2',  &lang), None);
        assert_eq!(Consonant.matches('f',  &lang), Some(true));
        assert_eq!(Consonant.matches('ы',  &lang), Some(false));
        assert_eq!(Consonant.matches('も', &lang), None);
        assert_eq!(Consonant.matches(';',  &lang), None);
        assert_eq!(Consonant.matches('.',  &lang), None);
        assert_eq!(Consonant.matches(',',  &lang), None);
        assert_eq!(Consonant.matches(' ',  &lang), None);
        assert_eq!(Consonant.matches('\0', &lang), None);
    }
}
