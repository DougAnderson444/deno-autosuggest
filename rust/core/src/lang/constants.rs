use super::CharClass;
use CharClass::{Consonant, Vowel};

pub const CHAR_CLASSES_LATIN: &[(CharClass, char)] = &[
    (Consonant, 'b'),
    (Consonant, 'c'),
    (Consonant, 'd'),
    (Consonant, 'f'),
    (Consonant, 'g'),
    (Consonant, 'h'),
    (Consonant, 'j'),
    (Consonant, 'k'),
    (Consonant, 'l'),
    (Consonant, 'm'),
    (Consonant, 'n'),
    (Consonant, 'p'),
    (Consonant, 'q'),
    (Consonant, 'r'),
    (Consonant, 's'),
    (Consonant, 't'),
    (Consonant, 'v'),
    (Consonant, 'w'),
    (Consonant, 'x'),
    (Consonant, 'z'),
    (Vowel, 'a'),
    (Vowel, 'e'),
    (Vowel, 'i'),
    (Vowel, 'o'),
    (Vowel, 'u'),
    (Vowel, 'y'),
];
