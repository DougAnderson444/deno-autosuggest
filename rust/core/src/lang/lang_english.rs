#![allow(dead_code)]

use rust_stemmers::{Algorithm, Stemmer};
use super::PartOfSpeech;
use super::Lang;

use PartOfSpeech::{
    Article,
    Preposition,
    Conjunction,
    Particle,
};


const FUNCTION_WORDS: &[(PartOfSpeech, &'static str)] = &[
    // Articles
    (Article, "a"),
    (Article, "an"),
    (Article, "the"),

    (Preposition, "at"),
    (Preposition, "by"),
    (Preposition, "for"),
    (Preposition, "from"),
    (Preposition, "in"),
    (Preposition, "of"),
    (Preposition, "on"),
    (Preposition, "to"),
    (Preposition, "since"),
    (Preposition, "before"),
    (Preposition, "till"),
    (Preposition, "untill"),
    (Preposition, "beside"),
    (Preposition, "under"),
    (Preposition, "below"),
    (Preposition, "over"),
    (Preposition, "above"),
    (Preposition, "across"),
    (Preposition, "through"),
    (Preposition, "into"),
    (Preposition, "towards"),
    (Preposition, "onto"),
    (Preposition, "off"),
    (Preposition, "out"),
    (Preposition, "about"),

    (Conjunction, "after"),
    (Conjunction, "although"),
    (Conjunction, "as"),
    (Conjunction, "because"),
    (Conjunction, "before"),
    (Conjunction, "but"),
    (Conjunction, "either"),
    (Conjunction, "for"),
    (Conjunction, "how"),
    (Conjunction, "if"),
    (Conjunction, "lest"),
    (Conjunction, "nor"),
    (Conjunction, "once"),
    (Conjunction, "once"),
    (Conjunction, "or"),
    (Conjunction, "since"),
    (Conjunction, "since"),
    (Conjunction, "since"),
    (Conjunction, "so"),
    (Conjunction, "than"),
    (Conjunction, "that"),
    (Conjunction, "that"),
    (Conjunction, "though"),
    (Conjunction, "till"),
    (Conjunction, "unless"),
    (Conjunction, "until"),
    (Conjunction, "until"),
    (Conjunction, "until"),
    (Conjunction, "what"),
    (Conjunction, "whatever"),
    (Conjunction, "when"),
    (Conjunction, "when"),
    (Conjunction, "whenever"),
    (Conjunction, "where"),
    (Conjunction, "whereas"),
    (Conjunction, "whereas"),
    (Conjunction, "wherever"),
    (Conjunction, "whether"),
    (Conjunction, "which"),
    (Conjunction, "whichever"),
    (Conjunction, "while"),
    (Conjunction, "while"),
    (Conjunction, "while"),
    (Conjunction, "whilst"),
    (Conjunction, "who"),
    (Conjunction, "whoever"),
    (Conjunction, "whom"),
    (Conjunction, "whomever"),
    (Conjunction, "whose"),
    (Conjunction, "why"),
    (Conjunction, "yet"),
    (Conjunction, "and"),
    // (Conjunction, "as if"),
    // (Conjunction, "as long as"),
    // (Conjunction, "as much as"),
    // (Conjunction, "as soon as"),
    // (Conjunction, "as though"),
    // (Conjunction, "assuming that"),
    // (Conjunction, "by the time"),
    // (Conjunction, "even if"),
    // (Conjunction, "even though"),
    // (Conjunction, "in case that"),
    // (Conjunction, "in case"),
    // (Conjunction, "in order that"),
    // (Conjunction, "in order"),
    // (Conjunction, "now that"),
    // (Conjunction, "only if"),
    // (Conjunction, "provided that"),
    // (Conjunction, "rather than"),
    // (Conjunction, "so that"),

    (Particle, "by"),
    (Particle, "in"),
    (Particle, "not"),
    (Particle, "on"),
    (Particle, "to"),
    (Particle, "oh"),
];

const UTF_COMPOSE_MAP: &[(&'static str, &'static str)] = &[];

const UTF_REDUCE_MAP: &[(&'static str, &'static str)] = &[];


pub fn lang_english() -> Lang {
    let mut lang = Lang::new();

    lang.set_stemmer(Some(Stemmer::create(Algorithm::English)));

    for (from, to) in UTF_COMPOSE_MAP { lang.add_unicode_composition(from, to); }
    for (from, to) in UTF_REDUCE_MAP  { lang.add_unicode_reduction(from, to); }

    for &(pos, word) in FUNCTION_WORDS { lang.add_pos(word, pos); }

    lang
}


#[cfg(test)]
mod tests {
    use crate::utils::to_vec;
    use super::PartOfSpeech;
    use super::{lang_english, UTF_COMPOSE_MAP, UTF_REDUCE_MAP};

    #[test]
    fn stem() {
        let lang = lang_english();
        let w    = to_vec("universe");
        assert_eq!(lang.stem(&w), 7);
    }

    #[test]
    fn get_pos() {
        let lang = lang_english();
        let w1   = to_vec("universe");
        let w2   = to_vec("the");
        assert_eq!(lang.get_pos(&w1), None);
        assert_eq!(lang.get_pos(&w2), Some(PartOfSpeech::Article));
    }

    #[test]
    fn unicode_compose() {
        let lang   = lang_english();
        let source = to_vec("universe");
        let norm   = lang.unicode_compose(&source);
        assert_eq!(norm, None);
    }

    #[test]
    fn unicode_reduce() {
        let lang   = lang_english();
        let source = to_vec("universe");
        let norm   = lang.unicode_reduce(&source);
        assert_eq!(norm, None);
    }

    #[test]
    fn unicode_compose_map_dimenstions() {
        for &(nfd, nfc) in UTF_COMPOSE_MAP {
            assert_eq!(nfd.chars().count(), 2);
            assert_eq!(nfc.chars().count(), 1);
        }
    }

    #[test]
    fn unicode_reduce_map_dimenstions() {
        for &(normal, reduced) in UTF_REDUCE_MAP {
            assert_eq!(normal .chars().count(), 1, "UTF_REDUCE_MAP['{}'] != 1", normal);
            assert_eq!(reduced.chars().count(), 1, "UTF_REDUCE_MAP['{}'].len() != 1", reduced);
        }
    }
}