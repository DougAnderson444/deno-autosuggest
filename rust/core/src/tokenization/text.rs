use std::fmt;
use crate::lang::Lang;
use super::CharPattern;
use super::Word;


#[derive(PartialEq)]
pub struct Text<T: AsRef<[char]>> {
    pub source: T,
    pub chars: T,
    pub words: Vec<Word>,
}


impl Text<Vec<char>> {
    pub fn from_vec(source: Vec<char>) -> Text<Vec<char>> {
        let len   = source.len();
        let chars = source.clone();
        Text {
            source,
            chars,
            words: vec![Word::new(len)],
        }
    }

    pub fn from_str(source: &str) -> Text<Vec<char>> {
        Text::from_vec(source.chars().collect())
    }
}


impl Text<Vec<char>> {
    pub fn to_ref<'a>(&'a self) -> Text<&'a [char]> {
        Text {
            source: &self.source,
            chars:  &self.chars,
            words:  self.words.clone(),
        }
    }
}


impl<'a> Text<&'a [char]> {
    pub fn to_own(&self) -> Text<Vec<char>> {
        Text {
            source: self.source.to_vec(),
            chars:  self.chars.to_vec(),
            words:  self.words.clone(),
        }
    }
}


impl<T: AsRef<[char]>> Text<T> {
    pub fn fin(mut self, fin: bool) -> Self {
        if let Some(word) = self.words.last_mut() {
            word.fin = fin;
        }
        self
    }

    pub fn is_empty(&self) -> bool {
        self.words.is_empty()
    }
}


impl Text<Vec<char>> {
    pub fn normalize(mut self, lang: &Option<Lang>) -> Self {
        if let Some(lang) = lang {
            if self.words.len() == 0 {
                return self;
            }
            if self.words.len() > 1 {
                panic!("Normalization should always be the first step");
            }

            if let Some(nfc) = lang.utf_compose(&self.source) {
                self.source           = nfc.clone();
                self.chars            = nfc;
                self.words[0].place.1 = self.chars.len();
            }

            if let Some((source, chars)) = lang.utf_reduce(&self.chars) {
                self.source           = source;
                self.chars            = chars;
                self.words[0].place.1 = self.chars.len();
            }
        }
        self
    }

    pub fn split<P: CharPattern>(mut self, pattern: &P) -> Self {
        let mut words = Vec::with_capacity(self.words.len());
        for word in &self.words {
            for splitted in word.split(&self.chars, pattern) {
                words.push(splitted);
            }
        }
        self.words = words;
        for (ix, word) in self.words.iter_mut().enumerate() {
            word.ix = ix;
        }
        self
    }

    pub fn strip<P: CharPattern>(mut self, pattern: &P) -> Self {
        for word in &mut self.words {
            word.strip(&self.chars, pattern);
        }
        self.words.retain(|w| w.len() > 0);
        for (ix, word) in self.words.iter_mut().enumerate() {
            word.ix = ix;
        }
        self
    }

    pub fn stem(mut self, lang: &Option<Lang>) -> Self {
        if let Some(lang) = lang {
            for word in &mut self.words {
                word.stem(&self.chars, lang);
            }
        }
        self
    }

    pub fn mark_pos(mut self, lang: &Option<Lang>) -> Self {
        if let Some(lang) = lang {
            for word in &mut self.words {
                word.mark_pos(&self.chars, lang);
            }
        }
        self
    }

    pub fn lower(mut self) -> Self {
        if self.chars.iter().any(|ch| ch.is_uppercase()) {
            for ch in &mut self.chars {
                *ch = ch.to_lowercase().next().unwrap_or(*ch);
            }
        }
        self
    }
}


impl<T: AsRef<[char]>> fmt::Debug for Text<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Text {{")?;
        for word in &self.words {
            let chars = word.view(self.chars.as_ref());
            write!(f, " \"")?;
            for (i, ch) in chars.iter().enumerate() {
                write!(f, "{}", ch)?;
                if i == word.stem - 1 && i != word.len() - 1 {
                    write!(f, "|")?;
                }
            }
            write!(f, "\"")?;
            if !word.fin { write!(f, "..")?; }
        }
        write!(f, " }}")?;
        Ok(())
    }
}


#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use crate::lang::{lang_english, lang_portuguese, lang_german};
    use super::{Word, Text};
    use super::super::{Chars, PartOfSpeech};

    use Chars::{
        Whitespaces,
        Punctuation,
    };

    #[test]
    fn text_normalize_nfd() {
        let text = Text::from_str("Conceição").normalize(&Some(lang_portuguese()));
        assert_debug_snapshot!((&text.source, &text.chars, &text.words[0]));
    }

    #[test]
    fn text_normalize_pad0() {
        let text = Text::from_str("straße").normalize(&Some(lang_german()));
        assert_debug_snapshot!((&text.source, &text.chars, &text.words[0]));
    }

    #[test]
    fn text_split() {
        let text = Text::from_str(" Foo Bar, Baz; ").split(&[Whitespaces, Punctuation]);
        assert_debug_snapshot!(text);
    }

    #[test]
    fn text_split_empty() {
        let text = Text::from_str(", ").split(&[Whitespaces, Punctuation]);
        assert_debug_snapshot!(text);
    }

    #[test]
    fn text_split_unfinished() {
        let text1 = Text::from_str(" Foo Bar, Baz"  ).fin(false).split(&[Whitespaces, Punctuation]);
        let text2 = Text::from_str(" Foo Bar, Baz; ").fin(false).split(&[Whitespaces, Punctuation]);
        assert_eq!(text1.words.last().unwrap().fin, false);
        assert_eq!(text2.words.last().unwrap().fin, true);
    }

    #[test]
    fn text_strip() {
        let chars = "-Foo- , Baz; ".chars().collect::<Vec<_>>();
        let text  = Text {
                source: chars.clone(),
                chars:  chars.clone(),
                words:  vec![
                    Word { ix: 0, place: (0,  5), stem: 5, pos: None, fin: true },  // "-Foo-"
                    Word { ix: 1, place: (6,  7), stem: 1, pos: None, fin: true },  // ","
                    Word { ix: 2, place: (8, 13), stem: 5, pos: None, fin: true },  // "Baz; "
                ],
            }
            .strip(&[Whitespaces, Punctuation]);
        assert_debug_snapshot!(text);
        assert_debug_snapshot!(text.words);
    }

    #[test]
    fn text_strip_unfinished() {
        let chars = "-Foo- Baz; ".chars().collect::<Vec<_>>();
        let text1 = Text {
                source: chars.clone(),
                chars:  chars.clone(),
                words:  vec![
                    Word { ix: 0, place: (0, 5), stem: 5, pos: None, fin: true },  // "-Foo-"
                    Word { ix: 1, place: (5, 8), stem: 3, pos: None, fin: true },  // "Baz"
                ],
            }
            .fin(false)
            .strip(&[Whitespaces, Punctuation]);

        let text2 = Text {
                source: chars.clone(),
                chars:  chars.clone(),
                words:  vec![
                    Word { ix: 0, place: (0,  5), stem: 5, pos: None, fin: true },  // "-Foo-"
                    Word { ix: 1, place: (5, 10), stem: 5, pos: None, fin: true },  // "Baz; "
                ],
            }
            .fin(false)
            .strip(&[Whitespaces, Punctuation]);

        assert_eq!(text1.words.last().unwrap().fin, false);
        assert_eq!(text2.words.last().unwrap().fin, true);
    }

    #[test]
    fn text_lower() {
        let chars = "Foo, Bar Baz".chars().collect::<Vec<_>>();
        let text  = Text {
                source: chars.clone(),
                chars:  chars.clone(),
                words:  vec![
                    Word { ix: 0, place: (0,  4), stem: 4, pos: None, fin: true }, // "Foo,"
                    Word { ix: 1, place: (5,  8), stem: 3, pos: None, fin: true }, // "Bar"
                    Word { ix: 2, place: (9, 12), stem: 3, pos: None, fin: true }, // "Baz"
                ],
            }
            .lower();
        assert_debug_snapshot!(text);
    }

    #[test]
    fn text_stem() {
        let chars = "hello universe".chars().collect::<Vec<_>>();
        let lang  = lang_english();
        let text  = Text {
                source: chars.clone(),
                chars:  chars.clone(),
                words: vec![
                    Word { ix: 0, place: (0,  5), stem: 5, pos: None, fin: true }, // "hello"
                    Word { ix: 1, place: (6, 14), stem: 8, pos: None, fin: true }, // "universe"
                ],
            }.stem(&Some(lang));
        assert_eq!(text.words[0].stem, 5);
        assert_eq!(text.words[1].stem, 7);
    }

    #[test]
    fn text_pos() {
        let chars = "the universe".chars().collect::<Vec<_>>();
        let lang  = lang_english();
        let text  = Text {
                source: chars.clone(),
                chars:  chars.clone(),
                words: vec![
                    Word { ix: 0, place: (0,  3), stem: 3, pos: None, fin: true }, // "hello"
                    Word { ix: 1, place: (4, 12), stem: 8, pos: None, fin: true }, // "universe"
                ],
            }.mark_pos(&Some(lang));
        assert_eq!(text.words[0].pos, Some(PartOfSpeech::Article));
        assert_eq!(text.words[1].pos, None);
    }
}
