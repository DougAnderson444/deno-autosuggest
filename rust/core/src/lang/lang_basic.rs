#![allow(dead_code)]

use super::Lang;
use super::constants::CHAR_CLASSES_LATIN;

pub fn lang_basic() -> Lang {
    let mut lang = Lang::new();
    for &(class, ch) in CHAR_CLASSES_LATIN { lang.add_char_class(ch, class); }
    lang
}
