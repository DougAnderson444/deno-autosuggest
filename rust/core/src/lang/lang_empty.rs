#![allow(dead_code)]

use rust_stemmers::{Algorithm, Stemmer};
use super::Lang;
use super::constants::CHAR_CLASSES_LATIN;

pub fn lang_empty() -> Lang {
    let mut lang = Lang::new();

    lang.set_stemmer(Some(Stemmer::create(Algorithm::English)));
    for &(class, ch) in CHAR_CLASSES_LATIN { lang.add_char_class(ch, class); }

    lang
}
