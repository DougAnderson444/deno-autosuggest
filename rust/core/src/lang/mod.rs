mod normalize;
mod lang_english;
mod lang_german;
mod lang_portuguese;
mod lang_russian;
mod lang_spanish;

pub use lang_german::lang_german;
pub use lang_english::lang_english;
pub use lang_spanish::lang_spanish;
pub use lang_portuguese::lang_portuguese;
pub use lang_russian::lang_russian;

use std::cell::RefCell;
use fnv::{FnvHashMap as HashMap};
use rust_stemmers::Stemmer;
use crate::utils::to_vec;
use crate::tokenization::PartOfSpeech;
use normalize::Normalize;


const BUFFER_CAPACITY: usize = 20;


pub struct Lang {
    stemmer:      Option<Stemmer>,
    pos_map:      HashMap<Vec<char>, PartOfSpeech>,
    compose_map:  HashMap<Vec<char>, Vec<char>>,
    reduce_map:   HashMap<Vec<char>, Vec<char>>,
    stem_buffer:  RefCell<String>,
    norm_buffer1: RefCell<Vec<char>>,
    norm_buffer2: RefCell<Vec<char>>,
}


impl Lang {
    pub fn new() -> Self {
        Self {
            stemmer:      None,
            pos_map:      HashMap::default(),
            compose_map:  HashMap::default(),
            reduce_map:   HashMap::default(),
            stem_buffer:  RefCell::new(String::with_capacity(BUFFER_CAPACITY)),
            norm_buffer1: RefCell::new(Vec::with_capacity(BUFFER_CAPACITY)),
            norm_buffer2: RefCell::new(Vec::with_capacity(BUFFER_CAPACITY)),
        }
    }

    pub fn set_stemmer(&mut self, stemmer: Option<Stemmer>) {
        self.stemmer = stemmer;
    }

    pub fn add_pos(&mut self, word: &str, pos: PartOfSpeech) {
        let source   = to_vec(word);
        let composed = self.unicode_compose(&source).unwrap_or(source);
        let reduced  = self.unicode_reduce(&composed);

        self.pos_map.insert(composed, pos);

        if let Some((_, reduced)) = reduced {
            self.pos_map.insert(reduced, pos);
        }
    }

    pub fn add_unicode_composition(&mut self, from: &str, to: &str) {
        self.compose_map.insert(to_vec(from), to_vec(to));
    }

    pub fn add_unicode_reduction(&mut self, from: &str, to: &str) {
        self.reduce_map.insert(to_vec(from), to_vec(to));
    }

    pub fn stem(&self, word: &[char]) -> usize {
        if let Some(stemmer) = &self.stemmer {
            let buffer = &mut *self.stem_buffer.borrow_mut();
            buffer.clear();
            buffer.extend(word.iter());
            let stem = stemmer.stem(&buffer);
            stem.chars().count()
        } else {
            word.len()
        }
    }

    pub fn get_pos(&self, word: &[char]) -> Option<PartOfSpeech> {
        self.pos_map.get(word).cloned()
    }

    pub fn unicode_compose(&self, word: &[char]) -> Option<Vec<char>> {
        let buffer = &mut *self.norm_buffer1.borrow_mut();
        buffer.clear();

        for (_, norm_chunk) in Normalize::new(word, &self.compose_map) {
            buffer.extend(norm_chunk);
        }

        if &buffer[..] == word {
            None
        } else {
            Some(buffer.clone())
        }
    }

    pub fn unicode_reduce(&self, word: &[char]) -> Option<(Vec<char>, Vec<char>)> {
        let buffer1 = &mut *self.norm_buffer1.borrow_mut();
        let buffer2 = &mut *self.norm_buffer2.borrow_mut();
        buffer1.clear();
        buffer2.clear();

        for (word_chunk, norm_chunk) in Normalize::new(word, &self.reduce_map) {
            buffer1.extend(word_chunk);
            buffer2.extend(norm_chunk);
            for _ in 0 .. norm_chunk.len() - word_chunk.len() {
                buffer1.push('\0');
            }
        }

        if &buffer2[..] == word {
            None
        } else {
            Some((buffer1.clone(), buffer2.clone()))
        }
    }
}


#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use crate::utils::to_vec;
    use crate::tokenization::PartOfSpeech;
    use super::Lang;

    fn get_lang() -> Lang {
        let mut lang = Lang::new();
        lang.add_unicode_composition("ó", "ó");
        lang.add_unicode_reduction("ó", "o");
        lang.add_unicode_reduction("õ", "oo");
        lang.add_pos("fóo", PartOfSpeech::Particle);
        lang
    }

    #[test]
    fn unicode_compose_nfc() {
        let input  = to_vec("foóbar");
        let output = get_lang().unicode_compose(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn unicode_compose_nfd() {
        let input  = to_vec("foóbar");
        let output = get_lang().unicode_compose(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn unicode_reduce_reduced() {
        let input  = to_vec("foobar");
        let output = get_lang().unicode_reduce(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn unicode_reduce_nfc() {
        let input  = to_vec("foóbar");
        let output = get_lang().unicode_reduce(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn unicode_reduce_fill0() {
        let input  = to_vec("fõbar");
        let output = get_lang().unicode_reduce(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn get_pos_source() {
        let input = to_vec("fóo");
        let pos   = get_lang().get_pos(&input);
        assert_eq!(pos, Some(PartOfSpeech::Particle));
    }

    #[test]
    fn get_pos_reduced() {
        let input = to_vec("foo");
        let pos   = get_lang().get_pos(&input);
        assert_eq!(pos, Some(PartOfSpeech::Particle));
    }

    #[test]
    fn get_pos_unknown() {
        let input = to_vec("bar");
        let pos   = get_lang().get_pos(&input);
        assert_eq!(pos, None);
    }
}