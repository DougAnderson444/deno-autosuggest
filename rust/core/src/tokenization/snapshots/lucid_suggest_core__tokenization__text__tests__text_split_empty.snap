---
source: src/tokenization/text.rs
expression: text
---
Text { }
